### PR TITLE
Add org.sil.FieldWorks

### PR DESCRIPTION
**Pull request by marksvc**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
  The app metainfo file passes other than that the validator doesn't seem to know about license "OFL-1.1-RFN".
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
  Permission usage is described in manifest comments.
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
  xchm/no-config.patch - The upstream project fixed the trouble I was experiencing with config files, but I can't upgrade to that newer xchm version until some other dependencies are sorted out.
  libxklavier/..Drop-intltool..patch - This is from Debian and helps build successfully.
  mono/mono-ping-path.patch - I haven't decided how to handle this with respect to upstream.

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: org.sil.FieldWorks.yml [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+        # Branch develop
+        tag: FieldWorks9.0.17
+        commit: 6d09370772ecd99f3efbe776c55c6561d1b7af88
+      - type: file
+        url: https://github.com/sillsdev/FwLocalizations/releases/download/v9.0.17-14-g37c0f8f/crowdin.zip
+        sha512: 08586f7479d5c99d668d4ec92a1824e02f2982683abf3d13549e07dc1378eaf1204547a1ce6cf9a29f4a9ee418aeaf558e00a13e25e9bd41a0de880a20de1f39
+        dest: fw/Downloads/
+      - type: shell
+        commands:
+          - mv fw/Downloads/crowdin.zip fw/Downloads/Crowdin.zip
```
Comment (hfiguiere) [inline org.sil.FieldWorks.yml:152]:
```suggestion
      dest-filename: Crowdin.zip
```

File: org.sil.FieldWorks.yml [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+              - |
+                tee nuget.config <<< '<?xml version="1.0" encoding="utf-8"?>
+                <configuration>
+                  <packageSources>
+                    <add key="LocalNugetPackages" value="'$(pwd)/nuget-sources'" />
+                  </packageSources>
+                  <disabledPackageSources>
+                    <add key="nuget.org" value="true" />
+                  </disabledPackageSources>
+                </configuration>'
```
Comment (hfiguiere) [inline org.sil.FieldWorks.yml:656]: you could use a "type: inline" to have this as an inline source.

File: org.sil.FieldWorks.yml [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+  # FieldWorks allows playing sound files and playing and recording in audio
+  # writing systems.
+  - --socket=pulseaudio
+  # FieldWorks can import language and dictionary data, restore from backup
+  # files, and link to pictures or other media to appear in a dictionary.
+  # Users need to save project data, project backups, Encoding Converter
+  # configurations, and configuration and settings of FieldWorks and included
+  # components.
+  # Users need to be able to export dictionaries to arbitrary locations.
+  - --filesystem=host
```
Comment (hfiguiere) [inline org.sil.FieldWorks.yml:47]: why would it need to write outside 'home' ?

File: flathub.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+  "only-arches": ["x86_64"]
```
Comment (hfiguiere) [inline flathub.json:2]: any specific reason? Mono works on aarch64

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by hfiguiere** [comment]
bot, build org.sil.FieldWorks

File: org.sil.FieldWorks.yml (hunk 3 above)
Comment (marksvc) [reply org.sil.FieldWorks.yml:47]: I was thinking if I limited it to home, users wouldn't be able to export to a USB Flash Drive.

File: flathub.json (hunk 4 above)
Comment (marksvc) [reply flathub.json:2]: I would be interested to try it some time. I know FieldWorks bundles a specific gecko for rendering html, and I don't think that includes aarch64 at present.

**Comment by marksvc** [comment]
The build failed building gtkmm2.4, but it did not appear to build its modules first.

File: org.sil.FieldWorks.yml (hunk 1 above)
Comment (marksvc) [reply org.sil.FieldWorks.yml]: Thank you :)

File: org.sil.FieldWorks.yml (hunk 2 above)
Comment (marksvc) [reply org.sil.FieldWorks.yml]: Thank you. I was not aware of inline.

**Comment by nedrichards** [comment]
bot, build org.sil.FieldWorks

**Comment by hfiguiere** [comment]
> The build failed building gtkmm2.4, but it did not appear to build its modules first.

gtk2 appear AFTER gtkmm2.4 from what I am seeing.

**Comment by marksvc** [comment]
But it's in the modules list, right? I don't see that the builder tried to build atkmm1.6 either.
```
- name: gtkmm2.4
  modules:
  - name: atkmm1.6
    ...
  - shared-modules/gtk2/gtk2.json
```

**Comment by marksvc** [comment]
Ok the problem is it couldn't deserialize modules. I'm checking on what happened with the submodule.

**Comment by hfiguiere** [comment]
> But it's in the modules list, right? I don't see that the builder tried to build atkmm1.6 either.
> 
> ```
> - name: gtkmm2.4
>   modules:
>   - name: atkmm1.6
>     ...
>   - shared-modules/gtk2/gtk2.json
> ```

yaml is basically unreadable. But I would put gtk2 FIRST in that list.

File: org.sil.FieldWorks.yml [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+# SOFTWARE.
+
+app-id: org.sil.FieldWorks
+
+# org.gnome.Platform provides some pieces that we would need to manually specify
+# if we use org.freedesktop.Platform.
+runtime: org.gnome.Platform
+# Target a runtime that is most similar to Ubuntu 20.04, for which we can easily
+# build and test on without flatpak.
+runtime-version: "3.36"
```
Comment (hfiguiere) [inline org.sil.FieldWorks.yml:30]:
```suggestion
runtime-version: "41"
```

3.36 is EOL.

**Comment by marksvc** [comment]
I think the newer runtime has resulted in some deprecation problems. Looking into it.

**Comment by marksvc** [comment]
Okay, the problems with not having gtk, and deprecation, should both be fixed.

**Comment by hfiguiere** [comment]
bot, build org.sil.FieldWorks

**Comment by marksvc** [comment]
The newer runtime is resulting in some compilation trouble. Looking.

**Comment by marksvc** [comment]
It will be good to fix the build failure that happens with the newer runtime, but I haven't learned its cause. I have set the runtime back to one where everything should build.

**Comment by barthalion** [comment]
Is the 3.36 runtime the last one that worked? There is also 3.38 and 40 that are slightly newer.

**Comment by marksvc** [comment]
@barthalion I checked again this morning. Runtime 3.38 and 40 both fail with the same problem. (And one needs to adjust the glibmm version back or forth to see.) I notice from some past work that mono 5 may not easily compile with newer versions of GCC. I see that org.gnome.Sdk 3.36 has GCC 9.3.0, while 3.38 and 40 have 10.2.0 and 41 has 11.2.0. So if that's the problem, then it may be a matter of getting GCC to allow for older code. (Although the error doesn't look like a _compilation error_. It's ... not clear what's gone wrong yet. So GCC may not be related to the trouble.)

**Comment by hfiguiere** [comment]
Not sure what glibmm issue you are having, but I found that in some app like Inkscape, adding 
```json
            "build-options": {
                "cxxflags": "-DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_66"
            },
```

helped if building failed because of glib.

**Comment by marksvc** [comment]
@hfiguiere Thank you. When I used newer org.gnome.Sdk v41, and older glibmm 2.64.2, the error in this situation was
```
glib/glibmm/binding.cc: In member function ‘void Glib::Binding::unbind()’:
glib/glibmm/binding.cc:131:27: error: ‘GObject* g_binding_get_source(GBinding*)’ is deprecated: Use 'g_binding_dup_source' instead [-Werror=deprecated-declarations]
  131 |   if (g_binding_get_source(gobj()))
      |       ~~~~~~~~~~~~~~~~~~~~^~~~~~~~
```
(And if I use an older org.gnome.Sdk with a newer glibmm, then I get an error like
> meson.build:3:0: ERROR: Meson version is 0.53.2 but project requires >= 0.54.0

:)

I can try a CXX flag like you showed if I need to do further convincing of the build to work with these various versions.

**Comment by marksvc** [comment]
`appstream-glib validate`'s list of licenses is not up to date, and it does not recognize the "OFL-1.1-RFN" license. I opened an issue for this at https://github.com/hughsie/appstream-glib/issues/435.

**Comment by barthalion** [comment]
bot, build org.sil.FieldWorks

**Comment by barthalion** [comment]
I don't think the license list will be really updated in a reasonable time, so please either drop the unknown license or try disabling the validation altogether like this: https://github.com/flathub/org.kde.katomic/blob/06a8d48244bcbdb113b18e1fffe9ad766d5b218e/flathub.json

**Comment by nedrichards** [comment]
bot, build org.sil.FieldWorks

**Comment by marksvc** [comment]
It had trouble fetching pieces of mono using git://. I'll look at how to work around that.

**Comment by hfiguiere** [comment]
> It had trouble fetching pieces of mono using git://. I'll look at how to work around that.

You should use `https://` for git repositories

**Comment by nedrichards** [comment]
> I don't think the license list will be really updated in a reasonable time, so please either drop the unknown license or try disabling the validation altogether like this: https://github.com/flathub/org.kde.katomic/blob/06a8d48244bcbdb113b18e1fffe9ad766d5b218e/flathub.json

Absolutely agreed. I got a [patch into appstream-glib master just now](https://github.com/hughsie/appstream-glib/commit/2e23a7780ea9c01be337d9bf05c7fdfbcd7d6ecc) but the line between there and 'trusted and tested code running on flathub' is not a clear or straight path.

**Comment by barthalion** [comment]
bot, build org.sil.FieldWorks

**Comment by barthalion** [comment]
Just to reiterate, it should be updated to the latest runtime eventually. I will merge it as-is for now.

/merge

**Pull request closed without merging**